File: chapter6/validateOrg.js
/**
 * Validate the organization data and employee data
 */
var jsonvalidate = require ("ujs-jsonvalidate");
var validate = jsonvalidate.validate;

/**
 * Validate organization structure and content.
 * @param {string} orgFile Organization JSON file name.
 * @param {string} empFile Employee JSON file name.
 */
function validateOrg (orgFile, empFile) {
    var orgSchema = "org_schema.json";
    var empSchema = "employee_schema.json";

    // validate organization data
    var units = null;
    validate (orgFile, orgSchema, null, null, function (code, data, msg) {
        if (code === jsonvalidate.VALID) {
            units = data.units;
        } else {
            console.log ("Error processing organization: " + msg);
            process.exit (code);
        }
    });

    // validate employee data
    var employees = null;
    validate (empFile, empSchema, null, null, function (code, data, msg) {
        if (code === jsonvalidate.VALID) {
            employees = data.employees;
        } else {
            console.log ("Error processing employees: " + msg);
            process.exit (code);
        }
    });

    // call custom validation functions
    verifyTopLevelUnit (units);
    verifyUniqueUnitIds (units);
    verifyUnitIds (units);
    verifyEmployeeUnits (employees, units);
}

/**
 * Verify org has one, and only one, top level unit.
 * @param {object[]} units Array of unit objects.
 */
function verifyTopLevelUnit (units) {
    var topLevelUnitsCount = 0;
    for (var ctr = 0; ctr < units.length; ctr ++) {
        if (units[ctr].unitOf === 0) {
            topLevelUnitsCount ++;
        }
    }
    // display results
    if (topLevelUnitsCount === 0) {
        console.log ("Error: Missing top level unit");
    } else if (topLevelUnitsCount === 1) {
        console.log ("Valid: Organization top level unit valid");
    } else {
        console.log ("Error: Multiple top level units defined");
    }
}

/**
 * Verify unitId is unique across all units. 
 * @param {object[]} units Array of unit objects.
 */
function verifyUniqueUnitIds (units) {
    for (var ctr1 = 0; ctr1 < units.length; ctr1 ++) {
        var currentUnitId = units[ctr1].unitId;
        for (var ctr2 = ctr1 + 1; ctr2 < units.length; ctr2 ++) {
            if (currentUnitId === units[ctr2].unitId) {
                console.log ("Error: Duplicate unitId " + currentUnitId);
                break;
            }
        }
    }
}

/**
 * Verify org has valid unitId for all unitOf references. 
 * @param {object[]} units Array of unit objects.
 */
function verifyUnitIds (units) {
    var orgValid = true;
    for (var ctr1 = 0; ctr1 < units.length; ctr1 ++) {
        var unitOf = units[ctr1].unitOf;
        if (unitOf !== 0) {
            var validUnitOf = false;
            for (var ctr2 = 0; ctr2 < units.length; ctr2 ++) {
                if (unitOf === units[ctr2].unitId) {
                    validUnitOf = true;
                    break;
                }
            }
            if (validUnitOf === false) {
                console.log ("Error: Invalid unitOf " + unitOf);
                orgValid = false;
            }
        }
    }
    if (orgValid === true) {
        console.log ("Valid: Organization hierarchy is valid");
    }
}

/**
 * Verify all employee unit references are valid org units.
 * @param {object[]} employees Array of employee objects.
 * @param {object[]} units Array of unit objects.
 */
function verifyEmployeeUnits (employees, units) {
    var allValid = true;
    for (var ctr1 = 0; ctr1 < employees.length; ctr1 ++) {
        var validUnit = false;
        var empUnit = employees[ctr1].unit;
        for (var ctr2 = 0; ctr2 < units.length; ctr2 ++) {
            if (empUnit === units[ctr2].unitId) {
                validUnit = true;
                break;
            }
        }
        if (validUnit === false) {
            console.log ("Error: Invalid employee unit " + empUnit);
            allValid = false;
        }
    }
    if (allValid === true) {
        console.log ("Valid: All employee unit references valid");
    }
}

// exports
exports.validateOrg = validateOrg;

File: chapter6/validateOrg.py
"""
 * Validate the organization data and employee data
"""
from jsonValidate import validate
import sys

def validateOrg (orgFile, empFile):
    """
    Validate organization structure and content.
    Args:
        orgFile Organization JSON file name.
        empFile Employee JSON file name.
    """
    orgSchema = "org_schema.json"
    empSchema = "employee_schema.json"

    # validate organization data
    units = None
    code, data, message = validate (orgFile, orgSchema, None, None)
    if code == 0:
        units = data["units"]
    else:
        print ("Error processing organization: " + message)
        sys.exit (code)

    # validate employee data
    employees = None
    code, data, message = validate (empFile, empSchema, None, None)
    if code == 0:
        employees = data["employees"]
    else: 
        print ("Error processing employees: " + message)
        sys.exit (code)

    verifyTopLevelUnit (units)
    verifyUniqueUnitIds (units)
    verifyUnitIds (units)
    verifyEmployeeUnits (employees, units)

def verifyTopLevelUnit (units):
    """ verify org has one, and only one, top level unit """
    topLevelUnitsCount = 0
    for unit in units:
        if unit["unitOf"] == 0:
            topLevelUnitsCount += 1

    # display results
    if topLevelUnitsCount == 0:
        print ("Error: Missing top level unit")
    elif topLevelUnitsCount == 1:
        print ("Valid: Organization top level unit valid")
    else:
        print ("Error: Multiple top level units defined")

def verifyUniqueUnitIds (units):
    """ verify unitId is unique across all units """
    for index1 in range (len (units)):
        currentUnitId = units[index1]["unitId"]
        for index2 in range (index1 + 1, len (units)):
            if currentUnitId == units[index2]["unitId"]:
                print ("Error: Duplicate unitId " + str (currentUnitId))
                break

def verifyUnitIds (units):
    """ verify org has valid unitId for all unitOf references """
    orgValid = True
    for units1 in units:
        if units1["unitOf"] != 0:
            validUnitOf = False
            for units2 in units:
                if units1["unitOf"] == units2["unitId"]:
                    validUnitOf = True
                    break

            if not validUnitOf:
                print ("Error: Invalid unitOf " + str (units1["unitOf"]))
                orgValid = False

    if orgValid:
        print ("Valid: Organization hierarchy is valid")

def verifyEmployeeUnits (employees, units):
    """ verify all employee unit references are valid org units """
    allValid = True
    for employee in employees:
        validUnit = False
        empUnit = employee["unit"]
        for unit in units:
            if empUnit == unit["unitId"]:
                validUnit = True
                break

        if not validUnit:
            print ("Error: Invalid employee unit " + str (empUnit))
            allValid = False

    if allValid:
        print ("Valid: All employee unit references valid")
